keps/sig-api-machinery/20190723-coordinated-discovery-document.md

```diff
@@ -0,0 +1,181 @@
+---
+title: Coordinated Discovery Documents for HA API Servers
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@yliaog"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-07-23
+last-updated: 2019-07-25
+status: provisional
+---
+
+# Coordinated Discovery Documents for HA API Servers
+
+## Table of Contents
+
+* [Problem Statement](#problem-statement)
+* [Summary of Proposed changes](#summary-of-proposed-changes)
+* [API changes](#api-changes)
+* [Changes to API servers](#changes-to-api-servers)
+   * [Calculating DiscoveryDocHashes](#calculating-discoverydochashes)
+   * [Periodically reporting DiscoveryDocHashes &amp; Garbage Collection](#periodically-reporting-discoverydochashes--garbage-collection)
+   * [Handling discovery requests](#handling-discovery-requests)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Graduation Plan](#graduation-plan)
+* [Alternatives](#alternatives)
+
+## Problem Statement
+
+API servers today compute discovery documents based on their local states. When
+HA API servers are rolling upgraded, API server instances can return
+different discovery documents to clients. This causes problems. For
+example, the storage migrator does not work reliably with HA API servers
+([details][]).
+
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+This KEP proposes a mechanism for HA API servers to expose to clients if
+all API server instances are serving consistent discovery document.
+
+## Summary of Proposed changes
+
+Periodically, each API server instance in an HA setup calculates a hash based on
+the discovery documents of the built-in APIs. The hash value is reported in a
+new API, `DiscoveryDocHashes`, together with the API server's ID. 
+
+When responding to discovery document requests, apart from the original
+discovery documents, the API server also sets a `Consistent` field to indicate if
+all API servers are serving the same discovery documents, i.e., if all hash
+values recorded in `DiscoveryDocHashes` are equal. 
+
+## API changes
+
+We introduce a new API `DiscoveryDocHashes`, in a new API group
+`APIServerCoordination.k8s.io`. 
+
+```golang
+type DiscoveryDocHashes struct {
```

Comment:
I suggest:
* Make the key to this the GVR; so we track separately per API
* The body is this:

```
Possibilities []PossibleDoc

...

type PossibleDoc struct {
  Hash string // key
  DocContents []byte // compressed; fits in etcd since it's just for a single GVR
  AcceptedBy []string // list of apiservers that "accept" this version of the doc
  ProposedBy []string // list of apiservers that "propose" this version, that is, will accept it as soon as they observer everyone proposing it.
}
```

When deciding what doc to show, apiserver gets the canonical list of apiservers from etcd (the endpoints record which we already manually curate). Then, for each resource, it looks at the above list. If all known apiservers propose a version, it will serve that one. Otherwise, it will serve the version which is accepted by the largest number of apiservers.

A background thread (the current doc assembly process) handles
* removing hashes that are not accepted or proposed by any live apiservers
* proposing new hashes when a change is observed
* once all apiservers have proposed the same version, modify to accept that one.
* we need this thread to work for the built in apis, too
* so, all of this code should run in the aggregator. We should not make any changes to the builtin, CRD, or AAs to support this.

Since the contents of the document is stored in etcd, all apiservers can serve the same document **even if they don't understand how to generate the document**.